### ISSUE TITLE
Add a grid search into cryojax?

### ISSUE THREAD

**Issue by mjo22** [mjo22/cryojax#201]
I am planning on writing a grid search so that I can do end-to-end parameter estimation in `jax`, and I am wondering if this has a place in `cryojax`. The plan is to write a very generic interface that can support different cost functions, algorithms, and ways of storing results.

A rough outline of the implementation I am imagining is the following:

1. *The grid:* The grid would be represented as an arbitrary pytree, whose leaves have leading "batch" dimensions. Namely, leaf $i$ is an array (or a pytree of arrays) with leading dimension(s) $N_i$. For $m$ leaves, the grid is then represented as a $N_1 \times N_2 \times ... \times N_m$ cartesian grid.
2. *The cost function:* The cost function $f$ would be a function that takes in a pytree of the same structure of the grid (and additional arguments), evaluated at grid point $(i_1, i_2, ..., i_m)$. This function can return a single value of the cost function, or it can return a *grid* of cost function evaluations. This would allow support for more clever grid searches than a simple exhaustive search, such as cross-correlation. In general, this sub-grid returned by $f$ would explore a region of parameter space unrelated to the $N_1 \times N_2 \times ... \times N_m$ grid.
3. *The solution:* There would need to be a flexible API for how the results of the grid search are stored. In a simple case, one could store the results of the best cost function evaluation at every grid point returned by the function $f$. In a more complicated case, one might want to use the grid search to marginalize away a portion of parameter space.

The question is: does this belong in `cryojax` or in a small separate library? In the meantime, will make a branch and work in `cryojax.inference`.

**Comment by mjo22** [comment]
Will be using the `optimistix` API and internals for inspiration: https://github.com/patrick-kidger/optimistix

**Comment by ehthiede** [comment]
Brings up an interesting point ... on one level we probably want some ability to optimize things out of the box, but too much could easily lead to feature-creep.  Maybe we can have an "optimize" entry-point with a default option or two, e.g. https://setuptools.pypa.io/en/latest/userguide/entry_point.html?  Grid search, or gradient descent could be the default option, which users could then overwrite using plugins... thoughts?

**Comment by mjo22** [comment]
Agreed on the feature-creep point. I am a little confused on how the plug-ins would work, but if I am understanding correctly, I would even go a step further and say we should make even less assumptions about how the user will want to perform optimization. 

To be more clear, let’s quickly compare against the `cryojax` recommendation for using gradient descent-based optimization: the recommendation is to write a script in the recommended pattern for some standard package (`optimistix`, `optax`, even `numpyro`), and use `cryojax` just as the simulation engine (and *maybe* also use it for a likelihood function). Again I am not sure I fully understand how the plug-ins would work, but my instinct is that the plug-ins would have to make some kind of assumption for how to use this model (?). Naturally then, the solution for the grid-search should be on the same footing as the gradient descent recommendation, since it’s just another way of doing optimization.

Issue is, I don’t think there is a standard package that would do the equivalent for a grid-search. So is the solution to add to `cryojax`? Create a small separate package (see my [attempt](https://github.com/patrick-kidger/optimistix/issues/55) to add this to `optimistix`)?

Either in `cryojax` or a separate package, I think there are ways we can avoid the feature-creep—for example (a) I’m imagining we do not provide utilities for any particular loss-function and (b) we don’t assume too much about what the user wants to do with each grid point (marginalize? find MAP estimate? etc). To address (b), we can write an `Abstract*` interface with one or two concrete implementations, that end-users can also subclass for themselves.

That said, this is what might kill the idea: if there is not a way to write an API that can generically encompass the ways in which one would want to parallelize the search, we very much run the risk of the feature-creep. We don’t want to be accepting feature requests for a host of different hyperparameters that determine how exactly the search loop is run. I think we would need to have a relatively fixed choice of how to run the loop and how to parallelize across devices. Again this could be solved with an `Abstract*` interface, but I don’t think this would really be appropriate.

Let me know what you think!

**Comment by mjo22** [comment]
Another point in all this: Michael, you’re basically just proposing to write a brute force loop, maybe that’s just too simple to add to the library!

I think that writing this in a generic way will require intermediate-to-advanced knowledge of jax.lax control flow and pytree manipulation, so I think there may be a lot of merit in making a nice API.

**Comment by mjo22** [comment]
Realized I forgot to address what you say about entry points for optimization, which is very important. What do you think about the model of documenting worked out jupyter notebook examples for this? Do you think this would be enough?

**Comment by mjo22** [comment]
Status report on this issue: for now, will continue to develop in the grid-search branch and work towards a jupyter notebook tutorial for the grid search. However, it does look likely that the grid search API belongs as a separate package.

**Comment by mjo22** [comment]
Closed with #207
